thorlcr/activities/indexread/thindexreadslave.cpp

```diff
@@ -719,10 +710,66 @@ class CIndexReadSlaveBase : public CSlaveActivity
     virtual void gatherActiveStats(CRuntimeStatisticCollection &activeStats) const
     {
         PARENT::gatherActiveStats(activeStats);
+        // *) jhtree stats will have been tracked as follows
+        //    - For superfiles, there will be 1 IContextLogger in contextLoggers[] for each subfile
+        //    - In all other cases, use 1st IContextLogger in contextLoggers[] for all parts
+        // *) File IO stats will have been tracked as follows
+        //    - the io stats will be tracked in keyManagers[]
+        //    - (There will be exactly 1 keyManager per part)
+        // Note: the KeyManagers[], contextLoggers[] and fileStats[] are not necessarilly the same size
+        // - KeyManagers[] will always be the same size as partDescs[]
+        // - contextLoggers[] will be the size of the number of subfiles for superfiles, otherwise 1
+        // - fileStats[] may be size of #subfiles for superfiles, size 1 for dynamic/variable files, or zero for everything else
+        // This function will:
+        // 1) In all cases, activeStats will have activity level jhtree & io stats
+        // 2) For superfiles and variable/dynamic files, fileStats[] will have jhtree & io stats
+        //    (If there are no fileStats, the file level stats can be extracted from activeStats)
+        if (partDescs.ordinality())
         {
-            CriticalBlock b(ioStatsCS);
-            if (lazyIFileIO)
-                mergeStats(activeStats, lazyIFileIO);
+            // reset required because within loop below, mergeStats() is used to build up stats for each file
+            for (auto & fileStatItem: fileStats)
+                fileStatItem->reset();
+            ISuperFileDescriptor *superFDesc = partDescs.item(0).queryOwner().querySuperFileDescriptor();
+            for (unsigned partNum=0; partNum<partDescs.ordinality(); partNum++)
+            {
+                IKeyManager * keyManager;
+                {
+                    CriticalBlock b(keyManagersCS);
+                    if (!keyManagers.isItem(partNum))
+                        continue;
+                    keyManager = &keyManagers.item(partNum);
+                }
+                if (fileStats.size()>0)
```

Comment:
Yes.  It can be zero even if partDescs>0.  fileStats is setup with CSlaveActivity::setupSpace4FileStats.   If it is not a superfile and it is not TIRvarfilename/TIRdynamicfilename, it will not create any entries in fileStats.  It will serialize the file stats in the activity stats and on the master, it will extract the file Stats from the activity stats.